torch/_inductor/pattern_matcher.py

```diff
@@ -854,6 +866,50 @@ def percolate_tags(node, recompute_tag, input_stops):
                         percolate_tags(new, old.meta["recompute"], args)
 
                     old.replace_all_uses_with(new)
+                    graph.erase_node(old)
+                    return
+
+                # `new` is not a node: it's a list of nodes.
+                #
+                # This happens when we want to replace a node that has a single
+                # packed return with multiple unpacked returns. We need to do
+                # some graph surgery here.
+                #
+                # Example:
+                #   def original_graph(x):
+                #      a = op(x)
+                #      b = a[0]
+                #      c = a[1]
+                #      ...
+                #
+                # Assume that we want to replace op(x) with the graph
+                #   def new_op(x):
+                #      w = x + 1
+                #      z = x + 2
+                #      return (w, z)
+                #
+                # We need to replace `op` with the contents of `new_op`,
+                # and then rewrite a[0] to be w and a[1] to be z, as so:
+                #   def new_graph(x):
+                #     w = x + 1
+                #     z = x + 2
+                #     b = w
+                #     c = z
+                #     ...
+                old_uses = list(old.users.keys())
+                for user in old_uses:
+                    idx = maybe_getitem(user)
+                    if idx is None:
+                        raise AssertionError("can't handle")
+                    replace(user, new[idx])
+                graph.erase_node(old)
+
```

Comment:
Would it be simpler to unpack the lists in `fwd_only` ?

Reply:
I don't think so? fwd_only looks like it's responsible for tracing the function that gets passed it. It has no notion of "here's the original graph and a new graph we want to replace a node with".

Reply:
We could feasibly be using aot autograd for tracing here instead of `make_fx` - (@bdhirsh suggested that once).  AOT_Autograd would do the unpacking right ?

Reply:
The problem is that the node we are trying to replace has a single packed return. We're trying to replace this with multiple returns. make_fx/AOTAutograd/etc will give us a function here that has unpacked returns.

We need to stitch up the unpacked returns back to the packed return via graph surgery.